rfd/0136-modern-signature-algorithms.md

```diff
@@ -0,0 +1,881 @@
+clients and hosts which have certificates issued by those CAs.
+
+This config will default to a `recommended` set of algorithms for each protocol
+chosen by us to balance security, compatibility, and performance.
+We will reserve the right to change this set of `recommended` algorithms when
+either:
+
+* the major version of the auth server's teleport.yaml config changes, or
+* in a major version release of Teleport.
+
+Most Teleport administrators will never need to see or interact with this config
+because they can trust that we will select a vetted set of standards-compliant
+algorithms that are trusted to be secure, and we will not break compatibility
+with internal Teleport components or third-party software unless deemed
+absolutely necessary for security reasons.
+
+Teleport administrators will be able to deviate from the `recommended`
+algorithms when they have a compliance need (they must use a particular
+algorithm) or a compatibility need (one of our selected algorithms is not
+supported by an external softare that interacts with Teleport in their
+deployment).
+
+Here is what the config will look like in its default state:
+
+```yaml
+ca_key_params:
+  user:
+    ssh:
+      algorithm: recommended
+      allowed_subject_algorithms: [recommended]
+    tls:
+      algorithm: recommended
+      allowed_subject_algorithms: [recommended]
+  host:
+    ssh:
+      algorithm: recommended
+      allowed_subject_algorithms: [recommended]
+    tls:
+      algorithm: recommended
+      allowed_subject_algorithms: [recommended]
+  db:
+    tls:
+      algorithm: recommended
+      allowed_subject_algorithms: [recommended]
+  openssh:
+    ssh:
+      algorithm: recommended
+      allowed_subject_algorithms: [recommended]
+  jwt:
+    jwt:
+      algorithm: recommended
+      allowed_subject_algorithms: [recommended]
+  saml_idp:
+    tls:
+      algorithm: recommended
+      allowed_subject_algorithms: [recommended]
+  oidc_idp:
+    jwt:
+      algorithm: recommended
+      allowed_subject_algorithms: [recommended]
+```
+
+You can imagine that if we had this config today, the `recommended` keyword here
+would expand to `RSA2048_PKCS1_SHA(256|512)` for all protocols.
+
+When we are ready to update the defaults (in a major config version or a major
+release) we will update the `recommended` rules to default to the following:
+
+(Note: there will be no actual change to the configuration resource which will
+still show `recommended`, the actual values will be computed within Teleport)
+
+(Reviewers: I need your help confirming if these selections are "good" or
+recommending alternatives)
+
+```yaml
+ca_key_params:
+  user:
+    ssh:
+      algorithm: Ed25519
+      # RSA2048 will initially be allowed for older `tsh` clients that don't
+      # know how to generate Ed25519 certs, and removed in a future major version
+      allowed_subject_algorithms: [Ed25519, RSA2048_PKCS1_SHA512]
```

Comment:
Yubikey PIV, which is used for [hardware key support](https://goteleport.com/docs/access-controls/guides/hardware-key-support/) only supports RSA and ECDSA signature algorithms. Currently we are using the ECDSA signature algorithm as I found it was orders of magnitude faster. In order to not break compatibility, would we also need to allow `ECDSA_P256_SHA256` here?

You mentioned this in your status update, so I'm guessing you are already aware, but I didn't see any note/comment on this yet.

Reply:
good point, added a mention